core/src/test/java/com/linecorp/armeria/client/retry/RetryingHttpClientTest.java

```diff
@@ -0,0 +1,264 @@
+
+        final Stopwatch sw = Stopwatch.createStarted();
+
+        final AggregatedHttpMessage res = client.execute(req).aggregate().get();
+        assertThat(res.content().toStringUtf8()).isEqualTo("Succeeded after retry");
+        assertThat(sw.elapsed(TimeUnit.MILLISECONDS)).isGreaterThanOrEqualTo(3000);
+    }
+
+    @Test
+    public void respectRetryAfterWithRfc1123() throws Exception {
+        final List<HttpStatus> status = Collections.singletonList(HttpStatus.SERVICE_UNAVAILABLE);
+        final RetryRequestStrategy<HttpRequest, HttpResponse> strategy = new RetryByStatus(status);
+        final HttpClient client = retryingClientWith(strategy, true);
+        final HttpHeaders req = HttpHeaders.of(HttpMethod.GET, "/retry-after-with-rfc-1123");
+
+        final Stopwatch sw = Stopwatch.createStarted();
+
+        final AggregatedHttpMessage res = client.execute(req).aggregate().get();
+        assertThat(res.content().toStringUtf8()).isEqualTo("Succeeded after retry");
+
+        // Since ZonedDateTime doesn't express exact time,
+        // just check out whether it is retried after delayed some time.
+        assertThat(sw.elapsed(TimeUnit.MILLISECONDS)).isGreaterThanOrEqualTo(1000);
+    }
+
+    private HttpClient retryingClientWith(RetryRequestStrategy<HttpRequest, HttpResponse> strategy) {
+        return new ClientBuilder(server.uri(SerializationFormat.NONE, "/"))
+                .decorator(HttpRequest.class, HttpResponse.class,
+                           RetryingHttpClient.newDecorator(strategy))
+                .build(HttpClient.class);
+    }
+
+    private HttpClient retryingClientWith(
+            RetryRequestStrategy<HttpRequest, HttpResponse> strategy, boolean useRetryAfter) {
+        return new ClientBuilder(server.uri(SerializationFormat.NONE, "/"))
+                .decorator(HttpRequest.class, HttpResponse.class,
+                           new RetryingHttpClientBuilder(strategy)
+                                   .backoffSupplier(() -> Backoff.withoutDelay().withMaxAttempts(5))
+                                   .useRetryAfter().newDecorator())
+                .build(HttpClient.class);
+    }
+
+    private static class RetryByContent implements RetryRequestStrategy<HttpRequest, HttpResponse> {
+        private final String retryContent;
+
+        RetryByContent(String retryContent) {
+            this.retryContent = retryContent;
+        }
+
+        @Override
+        public CompletableFuture<Boolean> shouldRetry(HttpRequest request, HttpResponse response) {
+            final CompletableFuture<AggregatedHttpMessage> future = response.aggregate();
+            return future.handle((message, thrown) ->
+                                         message != null &&
+                                         message.content().toStringUtf8().equalsIgnoreCase(retryContent)
+            );
+        }
+    }
+
+    private static class RetryByStatus implements RetryRequestStrategy<HttpRequest, HttpResponse> {
```

Comment:
I think the most of user might want to retry their requests depending on the response status of their request like this test case. So, how about move to the main code this kind of retry strategy? It would be useful to our users if we provide some pre-defined retry strategies.